Task 0

Klagefurt opened the pull request:
Исправил следующее:
1. Метод SerializeFiles: добавил проверку json файла перед записью - содержит ли он уже такой же экземпляр данного класса
2. Метод FindFactory: убрал обращение по индексу массива
3. Изменил чтение из excel файла на библиотеку EPPlus (вместо COM Interop)
4. Методы ReadFactories, ReadUnits, ReadTanks возвращают IEnumerable

Klagefurt committed:
Task 0

Klagefurt committed:
Исправил замечания

navferty commented on Reflection_Test/Reflection_Test/Student.cs:
```diff
@@ -0,0 +1,11 @@
+        private int _temp = 7;
+
+        public string Name { get; set; }
+        [My(Number = 11)]
```
Не очень понятно, за что отвечает атрибут `My`, а также что означает `Number`. Для исходного кода очень важно, чтобы имена сущностей (переменных, методов, свойств, классов, проектов и т.д.) максимально точно соответствовали их смыслу и тому, за что они отвечают.

Даже если это эксперименты, я бы предложил придумать немного более конкретный кейс: например, атрибут `AllowedRandeAttribute(int MinValue, int MaxValue)`, определяющий допустимый диапазон значений для свойства целочисленного типа.

Klagefurt replied:
Этот кусочек кода попал сюда по ошибке

navferty commented on dotnet-tasks/dotnet-tasks/Program.cs:
```diff
@@ -1,22 +1,36 @@
-        var tanks = GetTanks();
-        var units = GetUnits();
-        var factories = (Factory[])ReadFactoriesFromExcel();
+        bool readFromeExcel = true;
```
Чтобы не пришлось менять исходный код всякий раз, когда нужно тестировать разные варианты чтения, можно использовать аргументы командой строки - `string[] args`. Например, вызывая приложение с передачей аргументов `dotnet-tasks.exe --readFrom Excel`

Наряду с ручной реализацией чтения из аргументов, также предлагаю попробовать использовать вот этот микро-фреймворк для консольных приложений - [Cocona](https://github.com/mayuki/Cocona).

Klagefurt replied:
Попробовал использовать аргументы командной строки, в итоге оставил реализацию с помощью Cocona

navferty commented on dotnet-tasks/dotnet-tasks/Program.cs:
```diff
@@ -1,22 +1,36 @@
+        }
 
-        Console.WriteLine($"Количество резервуаров: {tanks.Length}, установок: {units.Length}");
+        Console.WriteLine($"Количество резервуаров: {tanks.Count()}, установок: {units.Count()}");
```
Следует обратить внимание, что `Count()` является методом, который вызывает итерацию по `IEnumerable`. Но дело в том, что этот интерфейс лишь означает, что по возвращённому значению можно проитерироваться в цикле `foreach` (или использовать методы, которые внутри основаны на этом цикле).

Но совершенно необязательно, что "под капотом" у объекта типа `IEnumerable` будет лежать именно коллекция. Это может быть некоторая абстракция, которая при каждой итерации цикла будет, например, читать очередную строку из файла, или каким-то ещё образом динамически генерировать очередное значение. Таким образом, множественная итерация по возвращённому из другого метода значению с типом `IEnumerable` может быть не самым желательным поведением.

Хотя этот интерфейс даёт возможность, например, реализовать потоковую обработку данных, которые за один раз могут даже не поместиться в оперативную память. Например, есть метод, который читает и возвращает поочерёдно каждую строку из csv файла (реализованный например как [метод-итератор](https://learn.microsoft.com/en-us/dotnet/csharp/iterators#enumeration-sources-with-iterator-methods), возвращающий значения через `yield return`, и который под капотом компилируется в [стейт-машину](https://csharpindepth.com/articles/IteratorBlockImplementation)). С другой стороны, его вызывает метод, который через `foreach` обрабатывает каждый элемент. При этом ни в один момент времени не приходится выделять память сразу под весь объём данных.

В данном случае, так как мы заранее можем быть уверены, что коллекция прочитана до конца и находится в памяти в материализованном виде, я бы предложил испрользовать интерфейс `IReadOnlyCollection`.

Klagefurt replied:
Спасибо за подробное объяснение. Использовал в итоге IReadOnlyCollection

navferty commented on dotnet-tasks/dotnet-tasks/Program.cs:
```diff
@@ -105,15 +168,21 @@ private static object[] ReadFactoriesFromExcel()
         foreach (var obj in array)
-        {    
+        {
             using (var streamwriter = File.AppendText(path))
```
Minor issue - можно использовать [using declaration](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/using#:~:text=also%20use%20a-,using%20declaration,-that%20doesn%27t%20require) вместо using statement. Это несколько уменьшит вложенность кода.

Klagefurt replied:
Исправил

navferty commented on dotnet-tasks/dotnet-tasks/Program.cs:
```diff
@@ -105,15 +168,21 @@ private static object[] ReadFactoriesFromExcel()
+
+                var jsonString = File.ReadAllText(path);
+                bool isUsed;
+                isUsed = jsonString.Contains(strObj);
```
А что если в файле лежат данные, но они отличаются от тех которые нужно записать? Возможно, вместо `AppendText` нужно использовать другой вариант, что не дописывать текст в конец файла (append), а перезаписать полностью содержимое файла?

Klagefurt replied:
Реализовал

navferty commented on dotnet-tasks/dotnet-tasks/Program.cs:
```diff
@@ -30,46 +44,95 @@ static void Main(string[] args)
-        try
+        ExcelPackage.LicenseContext = LicenseContext.NonCommercial;
+        var excelFileName = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Resources\\input.xlsx");
+        var file = new FileInfo(excelFileName);
```
Я бы предложил добавить на всякий случай проверку наличия файла перед попыткой чтения. Можно выбросить своё исключение: будет лучше получить исключение с исчерпывающим сообщением о том, что не найден эхель файл с заводами/установками/резервуарами, чем что-то малочитабельное в универсальном exception'е.

Klagefurt replied:
Добавил проверку в своем классе, унаследованном от Exception

navferty commented on dotnet-tasks/dotnet-tasks/Program.cs:
```diff
@@ -30,46 +44,95 @@ static void Main(string[] args)
+    private static IEnumerable<Factory> ReadFactories()
+    {
+        var factories = new List<Factory>();
+        var package = GetPackage();
```
В этих трёх методах происходит открытие файла с нуля. Но зачем каждый раз открывать заново файл с данными (который теоретически может быть достаточно объёмным), если можно продолжать читать его во всех трёх методах (например, заранее его открыть и передавать его через аргументы.

Klagefurt replied:
Исправил. В этом случае получилось, что package передается трем методам внутри одного using (var package = new ExcelPackage(file))

navferty commented on dotnet-tasks/dotnet-tasks/Program.cs:
```diff
@@ -30,46 +44,95 @@ static void Main(string[] args)
+    private static IEnumerable<Factory> ReadFactories()
+    {
+        var factories = new List<Factory>();
+        var package = GetPackage();
```
Насколько я вижу, класс `ExcelPackage` [реализует интерфейс `IDisposable`](https://epplussoftware.com/docs/5.3/api/OfficeOpenXml.ExcelPackage.html). Это значит, что после использования объект этого класса нужно обязательно очистить, вызвав метод `Dispose` (или завернув переменную в `using`).

Klagefurt replied:
Спасибо, забыл завернуть в using

navferty commented on dotnet-tasks/dotnet-tasks/Program.cs:
```diff
@@ -30,46 +44,95 @@ static void Main(string[] args)
+    {
+        var factories = new List<Factory>();
+        var package = GetPackage();
+        var worksheet = package.Workbook.Worksheets[0];
```
Я бы рекомендовал проверить, что в книге есть хотя бы один лист перед обращением по индексу (даже "нулевой" элемент в коллекции может отсутствовать, если она пустая).

Klagefurt replied:
Добавил проверку

navferty commented on dotnet-tasks/dotnet-tasks/Program.cs:
```diff
@@ -30,46 +44,95 @@ static void Main(string[] args)
         {
-            var excelFileName = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Resources\\input.xlsx");
+            Factory factory = new();
+            factory.ID = int.Parse(worksheet.Cells[row, col].Value.ToString());
```
Дважны происходит обращение к `worksheet.Cells[row, col].Value` и вызов `ToString()` на значении. Можно попробовать реорганизовав код (используя другой тип цикла, или вообще переписав на `while (true)` и инструкией `break` под условием), чтобы избежать лишних обращений к файлу.

Klagefurt replied:
Исправил

navferty commented on dotnet-tasks/dotnet-tasks/Program.cs:
```diff
@@ -30,46 +44,95 @@ static void Main(string[] args)
-            var excelFileName = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Resources\\input.xlsx");
+            Factory factory = new();
+            factory.ID = int.Parse(worksheet.Cells[row, col].Value.ToString());
+            factory.Name = worksheet.Cells[row, col + 1].Value.ToString();
```
Также рекомендую проверить, не будет ли выброшено исключение, если по этому индексу в книге уже нет заполненных строк и столбков (например, лист вообще абсолютно пустой без единой заполненной ячейки).

Klagefurt replied:
Добавил проверку

navferty commented on dotnet-tasks/dotnet-tasks/Program.cs:
```diff
@@ -30,46 +44,95 @@ static void Main(string[] args)
+    {
+        var units = new List<Unit>();
+        var package = GetPackage();
+        var worksheet = package.Workbook.Worksheets[1];
```
Аналогично по проверке наличия листа с нужным индексом. Лучше заранее проверить и выбросить исключение о недостающих листах в книге, чем разбираться с непонятными исключениями.

Klagefurt replied:
Добавил проверку

navferty commented on dotnet-tasks/dotnet-tasks/Program.cs:
```diff
@@ -1,22 +1,36 @@
-        var tanks = GetTanks();
-        var units = GetUnits();
-        var factories = (Factory[])ReadFactoriesFromExcel();
+        bool readFromeExcel = true;
```
Также можно добавить ещё опций для аргумента `readFrom`. Например, можно ещё реализовать чтение из csv файла(-ов), с использованием библиотеки [CsvHelper](https://joshclose.github.io/CsvHelper/examples/reading/get-class-records).

Klagefurt replied:
Использовал библиотеку CsvHelper. Добавил опцию --command readFromCSV в Cocona

navferty commented on dotnet-tasks/dotnet-tasks/Program.cs:
```diff
@@ -105,15 +168,21 @@ private static object[] ReadFactoriesFromExcel()
+        {
             using (var streamwriter = File.AppendText(path))
             {
                 var strObj = JsonConvert.SerializeObject(obj);
```
Предлагаю вместо `Newtonsoft.Json` использовать `System.Text.Json`, который в последние годы стал де-факто стандартным для новых .NET приложений.

Klagefurt replied:
Перешел на System.Text.Json

navferty commented on dotnet-tasks/dotnet-tasks/Program.cs:
```diff
@@ -30,46 +44,95 @@ static void Main(string[] args)
+        {
+            SerializeFiles(tanks);
+            SerializeFiles(units);
+            SerializeFiles(factories);
```
Возможно, вместо того чтобы пытаться использовать один и тот же файл (при этом неявно предполагая, что перед запуском программы его вовсе не существует), лучше явно записать все объекты структурированно? Например, сохранить все данные как объект с тремя свойствами, в каждом из которых лежит массив соответствующих объектов:

```
{
  "factories": [ { "name": "factory1", /*...*/ }, /*...*/ ],
  "units": [ /*...*/ ],
  "tanks": [ /*...*/ ]
}
```

Klagefurt replied:
Реализовал

Klagefurt committed:
Исправил замечания:
1. Добавил считывание csv файлов (CsvHelper)
2. Использовал Cocona
3. Вместо Newtonsoft.Json использовал System.Text.Json
4. Добавил сериаоизацию единым объектом
5. Добавил using для классов реализующих IDisposable
6. Добавил проверки try/catch для чтения excel файлов через EPPlus

Klagefurt committed:
Delete Reflection_Test directory
